src/test/ui/issues/issue-69306.stderr

```diff
@@ -8,8 +8,6 @@ LL |     const C: S0<u8> = Self(0);
    |
    = note: expected type parameter `T`
                         found type `{integer}`
-   = help: type parameters must be constrained to match other types
```

Comment:
maybe we should keep a `help` around that informs the user that a type parameter can never be equal to a concrete type, even if the caller would always choose that specific type?

Reply:
I would prefer for us to detect these kind of cases proactively and give a more targeted error. In the case of

```rust
struct S0<T>(T);
impl<T> S0<T> {
    const C: S0<u8> = Self(0);

    fn foo() {
        Self(0);
    }
}
```

we should come up with a more extensive explanation of what `Self` resolves to, something closer to

```
error[E0308]: mismatched types
  --> $DIR/issue-69306.rs:5:28
   |
LL | impl<T> S0<T> {
   |      - this type parameter
LL |     const C: S0<u8> = Self(0);
   |                       ---- ^ expected type parameter `T`, found integer
   |                       |
   |                       expected `T` because of this
   |
   = note: expected type parameter `T`
                        found type `{integer}`
note: `Self` resolves to `S0<T>`, which doesn't match `S0<{integer}>` becaue it is more general
   |
LL |     const C: S0<u8> = Self(0);
   |                       ^^^^

error[E0308]: mismatched types
  --> $DIR/issue-69306.rs:5:28
   |
LL | impl<T> S0<T> {
   |      - this type parameter
LL |     const C: S0<u8> = Self(0);
   |              ------   ^^^^^^^ expected `u8`, found type parameter `T`
   |              |
   |              expected `u8` because of this
   |
   = note: expected type parameter `T`
                        found type `{integer}`
note: `Self` resolves to `S0<T>`, which doesn't match `S0<u8>` becaue it is more general
   |
LL |     const C: S0<u8> = Self(0);
   |                       ^^^^

error[E0308]: mismatched types
  --> $DIR/issue-69306.rs:10:14
   |
LL | impl<T> S0<T> {
   |      - this type parameter
...
LL |         Self(0);
   |              ^ expected type parameter `T`, found integer
   |
   = note: expected type parameter `T`
                        found type `{integer}`
```